extensions/amp-story/1.0/amp-story-interactive-results.js

```diff
@@ -55,25 +60,118 @@ const buildResultsTemplate = (element) => {
-    if (e.option != null) {
-      categories[e.option.resultscategory] += 1;
+    if (e.type == InteractiveType.POLL) {
+      if (
+        e.option &&
+        e.option.resultscategory &&
+        categories[e.option.resultscategory] != null
+      ) {
+        categories[e.option.resultscategory] += 1;
+      }
     }
   });
-  return {
-    category: Object.keys(categories).reduce((a, b) =>
-      categories[a] >= categories[b] ? a : b
-    ),
-  };
+
+  // Returns category with most votes, first ones take precedence in ties
+  result.category = Object.keys(categories).reduce((a, b) =>
+    categories[a] >= categories[b] ? a : b
+  );
+  return result;
+};
+
+/**
+ * Processes the state and returns the condensed results for a percentage strategy
+ * @param {!Map<string, {option: ?./amp-story-interactive.OptionConfigType, interactiveId: string}>} interactiveState
+ * @param {?Array<!./amp-story-interactive.OptionConfigType>} options the attributes on the component
+ * @return {InteractiveResultsDef} the results
+ * @package
+ */
+export const processResultsPercentage = (interactiveState, options) => {
+  const result = {'percentage': null, 'category': null};
+
+  // Count quizzes and correct quizzes
+  let quizCount = 0;
+  let quizCorrect = 0;
+  Object.values(interactiveState).forEach((e) => {
+    if (e.type == InteractiveType.QUIZ) {
+      quizCount += 1;
+      if (e.option && e.option.correct != null) {
+        quizCorrect += 1;
+      }
+    }
+  });
+
+  // Percentage = (correct / total) but avoid divide by 0 error
+  result.percentage = quizCount == 0 ? 0 : 100 * (quizCorrect / quizCount);
+
+  // Get closest threshold that is lower than percentage, or lowest one if percentage is too low
+  let minThresholdDiff = -100;
+  options.forEach((option) => {
+    // ThresholdDiff is positive if it's lower than percentage (desired)
+    const currThresholdDiff =
+      result.percentage - parseFloat(option.resultsthreshold);
+    if (
+      // Curr meets the requirement and (is better or min doesnt meet)
+      (currThresholdDiff >= 0 &&
+        (minThresholdDiff > currThresholdDiff || minThresholdDiff < 0)) ||
+      // Curr doesnt meet the requirement, but min also doesnt and curr is better than min
+      (currThresholdDiff < 0 &&
+        minThresholdDiff < 0 &&
+        currThresholdDiff > minThresholdDiff)
+    ) {
+      result.category = option.resultscategory;
+      minThresholdDiff = currThresholdDiff;
+    }
+  });
+  return result;
+};
+
+/**
+ * Decides what strategy to use.
+ * If there are thresholds specified, it uses percentage; otherwise it uses category.
+ * @param {?Array<!./amp-story-interactive.OptionConfigType>} options the attributes on the component
+ * @return {string} the strategy
+ * @package
+ */
+export const decideStrategy = (options) => {
+  let strategy = 'category';
+  options.forEach((o) => {
+    if (o.resultsthreshold != null) {
+      strategy = 'percentage';
+    }
+  });
```

Comment:
Not that it really matters, but you can break the loop earlier using something like `options.some((o) => o.resultsthreshold !== null)`

Reply:
New favorite JS one-liner!